subprojects/core-api/core-api.gradle.kts

```diff
@@ -44,6 +44,10 @@ dependencies {
     testFixturesImplementation(project(":baseServices"))
 }
 
+classycle {
+    excludePatterns.set(listOf("org/gradle/**"))
```

Comment:
Hm...i don't quite understand this. Do we really want to skip those checks for all `core` and `core-api` classes?

Reply:
Seems like everything is connected there and we did not use the check there before. So effectively it is the same as before in these cases. We also can't easily resolve this (if we wanted) as a lot here is public API.

Reply:
I guess for the other projects Jendrik nailed down the exact problems, but for these two we had way too many, so we are not much better off than we were.

Would it be possible to explode this to just the packages that are problematic, even if it'd be a long list?

Reply:
> Would it be possible to explode this to just the packages that are problematic, even if it'd be a long list?

Would be possible of course. But for that you have to go deeper into it. There were cycles that span 10 or more packages. So to make this a useful list, I think you have to look into the code and understand what is causing the cycle and then probably add some comments to the list (or maybe even fix some cycles). Otherwise it would just be a long list where it is very unclear how the entries in the list relate to each other.

I would suggest, if we think it is useful to do that at some point, we do it as a separate task.

Reply:
(playing devils advocate) Do we really want this check given we kinda disable it for some of our most important pieces due to the large number of violations? Do we have plans to fix any of those? If we want static analysis, is it time to remove a dead tool (last release 2014) and use a more modern analyzer that does support our used languages and JVM versions?
Sorry Jendrik for opening that discussion, I think it diverts from your actual goal, I just wonder if the right approach for simplification is to drop this all together instead of streamlining it.

Reply:
I think we can discuss all of this if someone is willing to lead that discussion and possible also introduce a new tool. I think it makes makes most sense to do it at the moment you directly benefit from the tool. Like when you work in an area of the code base and want to improve the architecture there.

As I wrote, for now I just want to streamline things in our build and do not want to mix this up with more complex tasks/decisions.

You are right that removing would be another way of streamlining. But then we will loose a check if we do not introduce a new tool directly. I won't say in general that it is useless to use this tool. It may be "old", but it works. There are many parts of the code base where we use it right now.

I would like to merge this PR and then we can still have a discussion (sooner or later). Introducing a new tool afterwards should also be simpler as it is then more visible what the current situation is. And if the conclusion of that discussion is to remove this altogether, then that's also very easy to do.

Reply:
👍 to merging the PR as is.

Just FTR, having Classycle disabled in `:core` still means that if we move out some code from there to some other project, it will give us a chance to untangle any cycles. This has happened for example when I moved out projects like `:files`, `:persistent-cache`, ':execution` etc. (Unfortunately I haven't been able to untangle `:file-collections` yet.)

And yes, we absolutely should move to something more modern. I find that in general Classycle's error messages are only useful when you've just introduced a new cycle. If you want to enable it for existing code, well, good luck. I'm hoping that something like ArchUnit will give us better tools.